nexus/src/db/lookup.rs

```diff
@@ -0,0 +1,754 @@
+    fn lookup_for(
+        &self,
+        action: authz::Action,
+    ) -> BoxFuture<'_, LookupResult<Self::LookupType>>;
+}
+
+enum Key<'a, P> {
+    Name(P, &'a Name),
+    Id(LookupPath<'a>, Uuid),
+}
+
+impl<'a, T> GetLookupRoot for Key<'a, T>
+where
+    T: GetLookupRoot,
+{
+    fn lookup_root(&self) -> &LookupPath<'_> {
+        match self {
+            Key::Name(parent, _) => parent.lookup_root(),
+            Key::Id(lookup, _) => lookup,
+        }
+    }
+}
+
+impl<'a, P> Key<'a, P> {
+    fn lookup_type(&self) -> LookupType {
+        match self {
+            Key::Name(_, name) => LookupType::ByName(name.as_str().to_string()),
+            Key::Id(_, id) => LookupType::ById(*id),
+        }
+    }
+}
+
+pub struct LookupPath<'a> {
+    opctx: &'a OpContext,
+    datastore: &'a DataStore,
+}
+
+impl<'a> LookupPath<'a> {
+    pub fn new<'b, 'c>(
+        opctx: &'b OpContext,
+        datastore: &'c DataStore,
+    ) -> LookupPath<'a>
+    where
+        'b: 'a,
+        'c: 'a,
+    {
+        LookupPath { opctx, datastore }
+    }
+
+    pub fn organization_name<'b, 'c>(self, name: &'b Name) -> Organization<'c>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        Organization { key: Key::Name(self, name) }
+    }
+
+    pub fn organization_id(self, id: Uuid) -> Organization<'a> {
+        Organization { key: Key::Id(self, id) }
+    }
+
+    pub fn project_id(self, id: Uuid) -> Project<'a> {
+        Project { key: Key::Id(self, id) }
+    }
+
+    pub fn instance_id(self, id: Uuid) -> Instance<'a> {
+        Instance { key: Key::Id(self, id) }
+    }
+
+    pub fn disk_id(self, id: Uuid) -> Disk<'a> {
+        Disk { key: Key::Id(self, id) }
+    }
+}
+
+impl<'a> GetLookupRoot for LookupPath<'a> {
+    fn lookup_root(&self) -> &LookupPath<'_> {
+        self
+    }
+}
+
+impl<'a> Organization<'a> {
+    pub fn project_name<'b, 'c>(self, name: &'b Name) -> Project<'c>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        Project { key: Key::Name(self, name) }
+    }
+}
+
+impl<'a> Project<'a> {
+    pub fn disk_name<'b, 'c>(self, name: &'b Name) -> Disk<'c>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        Disk { key: Key::Name(self, name) }
+    }
+
+    pub fn instance_name<'b, 'c>(self, name: &'b Name) -> Instance<'c>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        Instance { key: Key::Name(self, name) }
+    }
+}
+
+macro_rules! define_lookup {
+    ($pc:ident) => {
+        paste::paste! {
+            pub struct $pc<'a> {
+                key: Key<'a, LookupPath<'a>>,
+            }
+
+            impl<'a> GetLookupRoot for $pc<'a> {
+                fn lookup_root(&self) -> &LookupPath<'_> {
+                    self.key.lookup_root()
+                }
+            }
+
+            // Do NOT make these functions public.  They should instead be
+            // wrapped by functions that perform authz checks.
+            async fn [<$pc:lower _lookup_by_id_no_authz>](
```

Comment:
you'll want to snake case this in the proc macro version (don't think there's any way to do that here)

Reply:
I think it is snake case -- the `:lower` lowercases whatever name you've passed into the macro.

Reply:
It's snake case in the existing cases because they are all one word. But `RouterRoute` will become `routerroute` instead of `router_route`, right?

Reply:
Sorry, by "this" I meant specifically `$pc`

Reply:
This is fixed in the new proc macro version.  Thanks for pointing it out -- I totally forgot there could be multiple words in the resource name 🤦